### PR TITLE
Default release version in the template cluster/nodepool subcommands

### PR DESCRIPTION
**Pull request by axbarsan**
As seen on https://gigantic.slack.com/archives/C02EVLE9W/p1637762623395000

The `--release` flag was accidentally made required by some internal change. This PR makes sure we always have a default value:
- for `template cluster`, we use the latest active release in the MC
- for `template nodepool`, we use the workload cluster's release version

**BONUS:** We now validate if the cluster exists when running `template nodepool`.

### CODE REVIEW

File: cmd/template/nodepool/runner.go [hunk 1]
```diff
@@ -123,3 +133,41 @@ func (r *runner) run(ctx context.Context, cmd *cobra.Command, args []string) err
...
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	switch wc := c.(type) {
+	case *cluster.Cluster:
+		return wc, nil
+	case *cluster.Collection:
+		targetWC := wc.Items[0]
+		return &targetWC, nil
```
Comment (axbarsan) [inline cmd/template/nodepool/runner.go:169]: Not sure if `&wc.Items[0]` would be safe here, as I believe `wc` would go out of scope and be garbage collected. Or would it be borrowed?

File: cmd/template/nodepool/runner.go [hunk 2]
```diff
@@ -85,12 +91,16 @@ func (r *runner) run(ctx context.Context, cmd *cobra.Command, args []string) err
...
-	if err != nil {
-		return microerror.Mask(err)
+		wc, err := getWorkloadCluster(ctx, c, config.Namespace, config.ClusterName, r.flag.Provider)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(config.ReleaseVersion) < 1 {
+			config.ReleaseVersion = wc.Cluster.Labels[label.ReleaseVersion]
+		} else {
+			config.ReleaseVersion = strings.TrimLeft(config.ReleaseVersion, "v")
```
Comment (anvddriesch) [inline cmd/template/nodepool/runner.go:102]: I'm not even sure we need an `else` case here. I can not think of any situation where we would want the nodepool to be a different version than the cluster.

File: cmd/template/nodepool/runner.go (hunk 1 above)
Comment (njuettner) [reply cmd/template/nodepool/runner.go:169]: maybe adding a test for this function would make sense, wdyt?

**Comment by whites11** [comment]
> **BONUS:** We now validate if the cluster exists when running `template nodepool`.

this means you cannot create the template for both a cluster and a node pool and create them with one single `kubectl apply` call?

File: cmd/template/nodepool/runner.go (hunk 2 above)
Comment (axbarsan) [reply cmd/template/nodepool/runner.go:102]: The `else` branch is used when the user provides the `--release` flag, and the value isn't defaulted.

If we want to remove the flag, it should probably be deprecated first for some time

**Comment by axbarsan** [comment]
> > **BONUS:** We now validate if the cluster exists when running `template nodepool`.
> 
> this means you cannot create the template for both a cluster and a node pool and create them with one single `kubectl apply` call?

Hmmm, good point. When there are multiple resource templates in a YAML, aren't they created in order, rather than all of them at the same time?

**Comment by whites11** [comment]
> > > **BONUS:** We now validate if the cluster exists when running `template nodepool`.
> > 
> > 
> > this means you cannot create the template for both a cluster and a node pool and create them with one single `kubectl apply` call?
> 
> Hmmm, good point. When there are multiple resource templates in a YAML, aren't they created in order, rather than all of them at the same time?

Yes, but If I run:

```
kubectl template cluster ...
kubectl template nodepool .. <fails because cluster is not applied
```

**Comment by axbarsan** [comment]
> > > > **BONUS:** We now validate if the cluster exists when running `template nodepool`.
> > > 
> > > 
> > > this means you cannot create the template for both a cluster and a node pool and create them with one single `kubectl apply` call?
> > 
> > 
> > Hmmm, good point. When there are multiple resource templates in a YAML, aren't they created in order, rather than all of them at the same time?
> 
> Yes, but If I run:
> 
> ```
> kubectl template cluster ...
> kubectl template nodepool .. <fails because cluster is not applied
> ```

Ah right. So we can change the implementation so the cluster would only be fetched if the release version is not provided.
My only concern with this is I'm not sure if it won't be confusing to only have the command fail if the `--release` flag is not provided and the WC does not exist.

File: cmd/template/nodepool/runner.go (hunk 1 above)
Comment (axbarsan) [reply cmd/template/nodepool/runner.go:166]: Added one now, please take a look

**Review by fiunchinho — CHANGES_REQUESTED** [review]
I know I proposed checking the cluster's release when creating a node pool but after thinking about it, I'm not sure it's the right approach anymore. 
By doing it this way, we prevent customers from generating their templates before applying them. Customers wouldn't be able to generate all their templates for both clusters and nodepools without having to apply the cluster template first.

I think this will make it harder for us to follow a gitops approach where all the manifests would be pushed to a git repository to be applied later on. I believe the release flag should be mandatory, or defaulted to the latest active release.

What do you think?

**Comment by axbarsan** [comment]
> I think this will make it harder for us to follow a gitops approach where all the manifests would be pushed to a git repository to be applied later on. I believe the release flag should be mandatory, or defaulted to the latest active release.

For clusters we already default to the latest active release. For node pools that wouldn't be possible, as there could be a mismatch between the node pool release and the cluster release.

We can make the flag mandatory, although it is a bit weird to have it like that for `template nodepool`, where it has no effect (execept for templating), but still have it optional for `template cluster`, where it actually has effect.

The way to go would probably be making the flag mandatory in both commands

**Comment by axbarsan** [comment]
> The fact the release is matching is checked by admission controllers. I would make it optional for both clusters and node pools and if the customer specifies two different versions it's their problem

We cannot leave it optional, as the commands determine based on it whether if it should template CAPI or GS resources. The user could template GS resources and the admission controller could default to a CAPI release.

**Comment by fiunchinho** [comment]
Then let's make it mandatory for both commands. What do you think @marians ?

**Comment by axbarsan** [comment]
Closed in favour of https://github.com/giantswarm/kubectl-gs/pull/589

**Pull request closed without merging**